plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,615 @@
+	fu_byte_array_append_uint8 (req, FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE);
+	fu_byte_array_append_uint32 (req, chk->address, G_LITTLE_ENDIAN);
+	fu_byte_array_append_uint32 (req, chk->data_sz, G_LITTLE_ENDIAN);
+	if (!fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					 req->data, req->len, error))
+		return FALSE;
+
+	/* check object create success or not */
+	if (!fu_pxi_device_wait_notify (self, 0x0, &opcode, NULL, error))
+		return FALSE;
+	if (opcode != FU_PXI_DEVICE_CMD_FW_OBJECT_CREATE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "FwWrite opcode invalid %02x",
+			     opcode);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_pxi_device_write_payload (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	g_autoptr(GByteArray) req = g_byte_array_new ();
+	fu_byte_array_append_uint8 (req, PXI_HID_DEV_OTA_OUTPUT_REPORT_ID);
+	g_byte_array_append (req, chk->data, chk->data_sz);
+	return fu_udev_device_pwrite_full (FU_UDEV_DEVICE (self), 0x0,
+					   req->data, req->len, error);
+}
+
+static gboolean
+fu_pxi_device_write_chunk (FuPxiDevice *self, FuChunk *chk, GError **error)
+{
+	guint32 prn = 0;
+	guint16 checksum = fu_pxi_device_calculate_checksum (chk->data, chk->data_sz);
+	guint16 checksum_tmp = 0;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* send create fw object command */
+	if (!fu_pxi_device_fw_object_create (self, chk, error))
+		return FALSE;
+
+	/* write payload */
+	chunks = fu_chunk_array_new (chk->data, chk->data_sz, chk->address,
+				     0x0, self->mtu_size);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk2 = g_ptr_array_index (chunks, i);
+		if (!fu_pxi_device_write_payload (self, chk2, error))
+			return FALSE;
+		prn++;
+		if (prn >= self->prn_threshold) {
+			guint8 opcode = 0;
+			if (!fu_pxi_device_wait_notify (self, 0x0, &opcode, &checksum_tmp, error))
+				return FALSE;
+			if (opcode != FU_PXI_DEVICE_CMD_FW_WRITE) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_READ,
+					     "FwWrite opcode invalid %02x",
+					     opcode);
+				return FALSE;
+			}
+			prn = 0;
+		}
+	}
+	self->checksum  +=  checksum;
+	/* check chunk is equal to FU_PXI_DEVICE_OBJECT_SIZE_MAX */ 
+	if (chk->data_sz == FU_PXI_DEVICE_OBJECT_SIZE_MAX) {
+		if (!fu_pxi_device_wait_notify (self, 0x0, NULL, &checksum_tmp, error))
+			return FALSE;
+	} else {
+		/* The last checksum from PRN */
+		if (self->checksum == checksum_tmp)
+			return TRUE;
+		else {
+			/* The last checksum from device fw length is zero notfiy */
+			if (!fu_pxi_device_wait_notify (self, 0x0, NULL, &checksum_tmp, error))
+				return FALSE;
+		}
+	}
+	g_debug ("checksum %x, table checksum %x", checksum_tmp, self->checksum);
+	if (checksum_tmp != self->checksum ) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_READ,
+			     "checksum fail, got %x, expected %x",
+			     checksum_tmp,
+			     checksum);
+		return FALSE;
+	}
```

Comment:
I'm a bit confused what all this is doing; can you explain please?

Reply:
wait notify condition in writing payload as below
1. RPN counter is over threshold
2. the length of writing chunk reach max object size (4096 byte)
3. the length of writing chunk reach fw length (last payload)

above condition could happen at same time. so I refactor the flow to avoid wait blocking situation

the code snippet in our deivce

```
//_fw_fast_write_size  is object size counter   
if ((_prn >= _prn_threshold) || (_fw_fast_write_size == 0) || (_fw_write_total >= _fw_length))
  {
        _prn = 0;
  
        if (_fw_fast_write_size == 0)
        {
            _last_object_offset += 1;
            DBG("_last_object_offset = %x\n", _last_object_offset);
        }
  
        ret.opcode = CMD_FW_WRITE;
        ret.crc = _crc;
        DBG("ret.crc = %x\n", ret.crc);
        ota_notify_cb(sizeof(ret), (uint8_t*)&ret); //notify to the host 
  }

```